pynvml/tests/test_smi.py

```diff
@@ -0,0 +1,77 @@
+from pynvml.smi import *
+import pytest
+import os
+
+# Fixture to initialize and finalize nvml
+@pytest.fixture(scope='module')
+def get_instance(request):
+    pytest.nvsmi = nvidia_smi.getInstance()
+
+@pytest.fixture
+def get_count(get_instance):
+    pytest.ngpus = pytest.nvsmi.DeviceQuery('count')['count']
+    print('['+str(pytest.ngpus)+' GPUs]', end =' ')
+    assert(pytest.ngpus>0)
+
+## ---------------------------- ##
+##       Start Query Tests      ##
+## ---------------------------- ##
+
+# Test free-memory query
+def test_query_memory(get_count):
+    def _free_mem_query(j):
+        return pytest.nvsmi.DeviceQuery('memory.free')['gpu'][j]['fb_memory_usage']['free']
+    for i in range(pytest.ngpus):
+        free_mem = _free_mem_query(i)
+        assert(free_mem >= 0)
+
+# Test gpu-utilization query
+def test_gpu_utilization(get_count):
+    def _gpu_util_query(j):
+        return pytest.nvsmi.DeviceQuery('utilization.gpu')['gpu'][j]['utilization']['gpu_util']
+    for i in range(pytest.ngpus):
+        gpu_util = _gpu_util_query(i)
+        assert(gpu_util >= 0)
+
+# Test memory-utilization query
+def test_memory_utilization(get_count):
+    def _memory_util_query(j):
+        return pytest.nvsmi.DeviceQuery('utilization.memory')['gpu'][j]['utilization']['memory_util']
+    for i in range(pytest.ngpus):
+        memory_util = _memory_util_query(i)
+        assert(memory_util >= 0)
+
+# Test tx memory-transfer utilization query
+def test_tx_util(get_count):
+    def _tx_util_query(j):
+        return pytest.nvsmi.DeviceQuery()['gpu'][j]['pci']['tx_util']
+    for i in range(pytest.ngpus):
+        tx_util = _tx_util_query(i)
+        assert(tx_util >= 0)
+
+# Test rx memory-transfer utilization query
+def test_rx_util(get_count):
+    def _rx_util_query(j):
+        return pytest.nvsmi.DeviceQuery()['gpu'][j]['pci']['rx_util']
+    for i in range(pytest.ngpus):
+        rx_util = _rx_util_query(i)
+        assert(rx_util >= 0)
```

Comment:
These two tests take the most time, I suspect because they grab all of the information and then filter it down, rather than just requesting the rx_util/tx_util information.  Is there a way to slim these down?

```
(ucx-dev) mrocklin@dgx16:~/pynvml$ py.test pynvml/tests/ -s -v --durations 10
=============================================================== test session starts ===============================================================
platform linux -- Python 3.7.3, pytest-4.4.1, py-1.8.0, pluggy-0.9.0 -- /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python
cachedir: .pytest_cache
rootdir: /home/nfs/mrocklin/pynvml
plugins: asyncio-0.10.0
collected 17 items

pynvml/tests/test_nvml.py::test_nvmlSystemGetNVMLVersion [NVML Version: 396.44] PASSED
pynvml/tests/test_nvml.py::test_nvmlSystemGetProcessName [Process: /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python] PASSED
pynvml/tests/test_nvml.py::test_nvmlSystemGetDriverVersion [Driver Version: 396.44] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetHandleBySerial [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetHandleByUUID [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetHandleByPciBusId [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetPowerUsage [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetMemoryInfo [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetUtilizationRates [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetPcieThroughput [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_query_memory [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_gpu_utilization [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_memory_utilization [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_tx_util [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_rx_util [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_pstate [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_temperature [8 GPUs] PASSED

============================================================ slowest 10 test durations ============================================================
12.61s call     pynvml/tests/test_smi.py::test_rx_util
12.18s call     pynvml/tests/test_smi.py::test_tx_util
0.49s call     pynvml/tests/test_nvml.py::test_nvmlDeviceGetPcieThroughput
0.26s call     pynvml/tests/test_smi.py::test_temperature
0.06s setup    pynvml/tests/test_smi.py::test_query_memory
0.05s setup    pynvml/tests/test_nvml.py::test_nvmlDeviceGetHandleBySerial
0.03s call     pynvml/tests/test_smi.py::test_query_memory
0.02s call     pynvml/tests/test_smi.py::test_gpu_utilization
0.02s setup    pynvml/tests/test_nvml.py::test_nvmlSystemGetNVMLVersion
0.02s call     pynvml/tests/test_nvml.py::test_nvmlDeviceGetPowerUsage
```